File: examples/camera_ir_do_ffc_control.cpp
/**
 * This sample will show you how to control the FFC in the thermal camera of the payload
 * 1. Set/Get FFC mode: Auto / Manual
 * 2. Trigger FFC, you will hear the "click" from the camera 
 **/

#include "stdio.h"
#include"payloadSdkInterface.h"

PayloadSdkInterface* my_payload = nullptr;
bool time_to_exit = false;

#if (CONTROL_METHOD == CONTROL_UART)
T_ConnInfo s_conn = {
    CONTROL_UART,
    payload_uart_port,
    payload_uart_baud
};
#else
T_ConnInfo s_conn = {
    CONTROL_UDP,
    udp_ip_target,
    udp_port_target
};
#endif

void quit_handler(int sig);

int ffc_trigger_cnt_max=5;
int ffc_trigger_cnt = 0;

int main(int argc, char *argv[]){
	printf("Starting IR FFC Control example...\n");
	signal(SIGINT,quit_handler);

	// creat payloadsdk object
	my_payload = new PayloadSdkInterface(s_conn);

	// init payload
	my_payload->sdkInitConnection();
	printf("Waiting for payload signal! \n");

	// check connection
	my_payload->checkPayloadConnection();

	printf("[IR] Forcing view source to IR...\n");
	my_payload->setPayloadCameraParam(PAYLOAD_CAMERA_VIEW_SRC, PAYLOAD_CAMERA_VIEW_IR, PARAM_TYPE_UINT32);

	// change FFC mode to Auto
	my_payload->setPayloadCameraFFCMode(FFC_MODE_AUTO);
	SDK_LOG("Change FFC to Auto, waiting for 5secs.");
	usleep(5000000);


	// change FFC mode to Manual
	my_payload->setPayloadCameraFFCMode(FFC_MODE_MANUAL);
	SDK_LOG("Change FFC to Manual, waiting for 5secs.");
	usleep(5000000);
	
	// check payload messages
	while(1){
		if(ffc_trigger_cnt < ffc_trigger_cnt_max){
			ffc_trigger_cnt++;
			SDK_LOG("Trigger FFC, %d", ffc_trigger_cnt);
			my_payload->setPayloadCameraFFCTrigg();
			usleep(3000000);
		}
		else{
			break;
		}
	}

	SDK_LOG("Done. Exit...");
	return 0;
}

void quit_handler( int sig ){
    printf("\n");
    printf("TERMINATING AT USER REQUEST \n");
    printf("\n");

    // close payload interface
    try {
        my_payload->sdkQuit();
    }
    catch (int error){}

    // end program here
    exit(0);
}

File: examples/camera_ir_do_ffc_control.py
'''
 * This sample will show you how to control the FFC in the thermal camera of the payload
 * 1. Set/Get FFC mode: Auto / Manual
 * 2. Trigger FFC, you will hear the "click" from the camera 
'''

#!/usr/bin/env python3
import sys
import os

# Add the libs directory to the path
sys.path.append(os.path.join(os.path.dirname(__file__), '..', 'libs'))

# Import config first to setup environment automatically
from config import config

import time
import signal
import sys
from payload_sdk import PayloadSdkInterface, ffc_mode_t
from payload_define import *

my_payload = None
time_to_exit = False

# Counter for FFC trigger
ffc_trigger_cnt_max = 5
ffc_trigger_cnt = 0

# Signal handler for quitting
def quit_handler(sig, frame):
    global my_payload, time_to_exit
    print("\nTERMINATING AT USER REQUEST")
    time_to_exit = True

    # Close payload interface
    if my_payload:
        try:
            my_payload.sdkQuit()
        except Exception as e:
            print(f"Error while quitting payload: {e}")
    # End program    
    sys.exit(0)

def main():
    global my_payload, ffc_trigger_cnt, time_to_exit

    print("Starting IR FFC Control example...\n")
    signal.signal(signal.SIGINT, quit_handler)

    # Create payloadsdk object
    my_payload = PayloadSdkInterface()

    # Init payload
    my_payload.sdkInitConnection()
    print("Waiting for payload signal!\n")

    # Check connection
    my_payload.checkPayloadConnection()

    # Change FFC mode to Auto
    my_payload.setPayloadCameraFFCMode(ffc_mode_t.FFC_MODE_AUTO)
    print("Change FFC to Auto, waiting for 5secs.")
    time.sleep(5)

    # Change FFC mode to Manual
    my_payload.setPayloadCameraFFCMode(ffc_mode_t.FFC_MODE_MANUAL)
    print("Change FFC to Manual, waiting for 5secs.")
    time.sleep(5)  

    # Perform FFC trigger operations in a loop
    while not time_to_exit:
        if ffc_trigger_cnt < ffc_trigger_cnt_max:
            ffc_trigger_cnt += 1
            print(f"Trigger FFC, {ffc_trigger_cnt}")
            my_payload.setPayloadCameraFFCTrigg()
            time.sleep(3)
        else:
            break

    print("Done. Exit...")

    # Close payload interface
    try:
        my_payload.sdkQuit()
        print("Payload connection closed successfully.")
    except Exception as e:
        print(f"Error while quitting payload: {e}")    

if __name__ == "__main__":
    main()